rasp/base.py

```diff
@@ -8,53 +8,43 @@
 
 class Engine(object):
     def get_page_source(self, url):
-        raise EngineError("get_page_source not implemented")
+        raise NotImplemented("get_page_source not implemented for {}"
+                             .format(str(self.__class__.__name__)))
 
     def cleanup(self):
         return
 
-    def clone(self):
-        raise EngineError("clone not implemented")
-
 
 class DefaultEngine(Engine):
     def __init__(self, data=None, headers=None):
         self.data = data
+        self.headers = headers or {'User-Agent': DEFAULT_USER_AGENT}
 
-        if headers is None:
-            self.headers = {'User-Agent': DEFAULT_USER_AGENT}
-        else:
-            self.headers = headers
-
-        return
+    def __copy__(self):
+        return DefaultEngine(self.data, self.headers)
 
-    def get_page_source(self, url):
+    def get_page_source(self, url, data=None):
+        if not url:
+            return EngineError('url needs to be specified')
```

Comment:
It's a much better pattern to test for the negative instead of having an unnecessary increase in indentation for the rest of the positive condition code.